### ISSUE TITLE
pathlib.Path.iterdir doesn't raise an exception until you start iterating

### ISSUE THREAD

**Issue by 0c97e840-ae8b-4fa9-b95e-fe2331d61b5b** [python/cpython#78722]
BPO | [34541](https://bugs.python.org/issue34541)
--- | :---
Nosy | @pitrou, @prudvinit
PRs | <li>python/cpython#8996</li><li>python/cpython#8999</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2018-08-29.08:20:21.120>
labels = ['3.8', 'type-feature', 'library']
title = "pathlib.Path.iterdir doesn't throw an exception until you start iterating"
updated_at = <Date 2019-05-23.22:47:36.119>
user = 'https://bugs.python.org/PaulPinterits'
```

bugs.python.org fields:
```python
activity = <Date 2019-05-23.22:47:36.119>
actor = 'cheryl.sabella'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2018-08-29.08:20:21.120>
creator = 'Paul Pinterits'
dependencies = []
files = []
hgrepos = []
issue_num = 34541
keywords = ['patch']
message_count = 3.0
messages = ['324307', '324332', '324334']
nosy_count = 3.0
nosy_names = ['pitrou', 'Paul Pinterits', 'prudvinit']
pr_nums = ['8996', '8999']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue34541'
versions = ['Python 3.8']
```

</p></details>

### Linked PRs
* gh-107320

**Comment by barneygale** [comment]
Work-around: call `list(path.iterdir())`:

```python
try:
    children = list(path_that_doesnt_exist.iterdir())
except FileNotFoundError:
    print("directory doesn't exist")
for path in children:
    print(path.stat().st_size)
```

I don't think we can change the `iterdir()` behaviour at this stage, nor does it justify the introduction of a similar method like `listdir()` to my mind.

**Comment by CAM-Gerlach** [comment]
Yeah, the fact that it returns a generator/lazy iterator seems pretty fundamental to `iterdir`'s behavior, and matches the other current and past `iter*` functions/methods in Pythons, as well as pathlib.glob` and similar methods (as well as builtins like `range` and others) which would make it suddenly behaving eagerly quite surprising to many/most users, as well as sacrificing performance/efficiency. `list(<lazy iterator>)` to make something iterable is a common idiom in Python to handle this situation (and others), and adding a separate method to do the same thing eagerly seems thoroughly redundant.

**Comment by barneygale** [comment]
I'm going to close this issue per CAM's rationale above. Use `list(path.iterdir())` as a workaround. Thanks all.

**Comment by jaraco** [comment]
I don't believe `list(iterdir())` is a suitable workaround, as it requires the entire directory to be materialized in memory, increasing storage with little benefit. I'd argue that itertools (or similar) should provide a mechanism to filter or stop on exceptions, so consuming the iterable can continue to be lazy.

[more_itertools.iter_except](https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.iter_except) almost handles this case, but (a) requires the callable to be passed and (b) can't handle `StopIteration` as a stop condition.

**Comment by barneygale** [comment]
Upon reflection I agree. I've put a PR up. Do you have a preference on whether it should be backported?

**Comment by barneygale** [comment]
A [comment](https://github.com/python/cpython/pull/106337#discussion_r1310991385) from @AA-Turner:

> I agree fixing iterdir at the source would be cleaner

**Comment by barneygale** [comment]
Fixed in 3.13 / #107320 / [bdc3c88](https://github.com/python/cpython/commit/bdc3c884cdc90102ad68b6b55dc9b988e729ae35)

**Comment by jaraco** [comment]
Thanks for the fix!

> Upon reflection I agree. I've put a PR up. Do you have a preference on whether it should be backported?

This feels like a change in behavior / improvement, so probably not suitable for a backport.
